client/trino-jdbc/src/main/java/io/trino/jdbc/TrinoResultSet.java

```diff
@@ -174,22 +187,46 @@ public AsyncIterator(Iterator<T> dataIterator, StatementClient client)
                 }
                 finally {
                     semaphore.release();
+                    finished = true;
                 }
-            }, executorService);
+            });
         }
 
         public void cancel()
         {
             future.cancel(true);
+            cancelled = true;
+            cleanup();
         }
 
         public void interrupt(InterruptedException e)
         {
-            client.close();
+            cleanup();
             Thread.currentThread().interrupt();
             throw new RuntimeException(new SQLException("ResultSet thread was interrupted", e));
         }
 
+        private void cleanup()
+        {
+            // When thread interruption is mis-handled by underlying implementation of `client`, the thread which
+            // is working for `future` may be blocked by `rowQueue.put` (`rowQueue` is full) and will never finish
```

Comment:
I think being blocked could also happen without mis-handling of the interrupts, right?

Reply:
Block happens, but will be interrupted if it is properly handled.